cpp/src/hash/hashing.cu

```diff
@@ -0,0 +1,373 @@
+  auto block_partition_sizes = rmm::device_vector<size_type>(grid_size * num_partitions);
+  auto global_partition_sizes = rmm::device_vector<size_type>(num_partitions);
+  CUDA_TRY(cudaMemsetAsync(global_partition_sizes.data().get(), 0, num_partitions * sizeof(size_type), stream));
+
+  // If the number of partitions is a power of two, we can compute the partition 
+  // number of each row more efficiently with bitwise operations
+  if (is_power_two(num_partitions)) {
+    // Determines how the mapping between hash value and partition number is computed
+    using partitioner_type = bitwise_partitioner<hash_value_type>;
+
+    // Computes which partition each row belongs to by hashing the row and performing
+    // a partitioning operator on the hash value. Also computes the number of
+    // rows in each partition both for each thread block as well as across all blocks
+    compute_row_partition_numbers<has_nulls>
+        <<<grid_size, block_size, num_partitions * sizeof(size_type), stream>>>(
+            *device_input, num_rows, num_partitions,
+            partitioner_type(num_partitions), row_partition_numbers.data().get(),
+            block_partition_sizes.data().get(), global_partition_sizes.data().get());
+
+  } else {
+    // Determines how the mapping between hash value and partition number is computed
+    using partitioner_type = modulo_partitioner<hash_value_type>;
+
+    // Computes which partition each row belongs to by hashing the row and performing
+    // a partitioning operator on the hash value. Also computes the number of
+    // rows in each partition both for each thread block as well as across all blocks
+    compute_row_partition_numbers<has_nulls>
+        <<<grid_size, block_size, num_partitions * sizeof(size_type), stream>>>(
+            *device_input, num_rows, num_partitions,
+            partitioner_type(num_partitions), row_partition_numbers.data().get(),
+            block_partition_sizes.data().get(), global_partition_sizes.data().get());
+  }
+
+  // Compute in-place exclusive scan of all blocks' partition sizes to determine 
+  // the starting point for each blocks portion of each partition in the output
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         block_partition_sizes.begin(),
+                         block_partition_sizes.end(),
+                         block_partition_sizes.begin());
+
+  // Compute in-place exclusive scan of size of each partition to determine
+  // offset location of each partition in final output.
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         global_partition_sizes.begin(),
+                         global_partition_sizes.end(),
+                         global_partition_sizes.begin());
+
+  // Copy the result of the exlusive scan to the output offsets array
+  // to indicate the starting point for each partition in the output
+  std::vector<size_type> partition_offsets(num_partitions);
```

Comment:
Use a `thrust::host_vector` for automatic device to host copying.

Reply:
Ah good idea. This doesn't take a stream though, so I assuming it's synchronizing on the default stream?

Reply:
Actually... I'm not sure that I need `partition_offsets` in host memory in the new implementation, so this might be a moot point.